### PR TITLE
Implement a mechanism to have ad networks specify client ids that get…

### LINKED ISSUE #1324: Implement feature to pass CID to ads
This helps ads that otherwise use 1p cookies for things like frequency capping.

### PR DESCRIPTION
**Pull request by cramforce**
This is not currently used anywhere.

Fixes #1324

### CODE REVIEW

**Comment by cramforce** [comment]
This is now ready for review.

File: builtins/amp-ad.js [hunk 1]
```diff
@@ -202,28 +203,56 @@ export function installAd(win) {
...
+      const scope = clientIdScope[this.element.getAttribute('type')];
+      if (!scope) {
+        return Promise.resolve();
+      }
+      return cidForOrNull(this.getWin()).then(cidService => {
+        if (!cidService) {
+          return Promise.resolve();
+        }
+        return cidService.get(scope,
+            // TODO(@cramforce): Hook up user notification service.
```
Comment (erwinmombay) [inline builtins/amp-ad.js]: anything else needed to hook it up?

Comment (cramforce) [inline builtins/amp-ad.js]: A resolution on how to do #1313

File: src/cid.js [hunk 2]
```diff
@@ -27,3 +27,11 @@ import {getElementService} from './service';
 export function cidFor(window) {
   return getElementService(window, 'cid', 'amp-analytics');
 };
+
+/**
+ * @param {!Window} window
+ * @return {!Promise<!Cid>}
```
Comment (erwinmombay) [inline src/cid.js]: since it can be null, shouldn't it be `!Promise<?Cid>`

File: docs/create_page.md [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# How to Create a Basic AMP HTML Page
+ifram# How to Create a Basic AMP HTML Page
```
Comment (erwinmombay) [inline docs/create_page.md]: typo?

File: ads/README.md [hunk 4]
```diff
@@ -34,9 +34,15 @@ We will provide the following information to the ad:
   In browsers that support `location.ancestorOrigins` you can trust that the `origin` of the
   location is actually correct (So rogue pages cannot claim they represent an origin they do not actually represent).
 - `window.context.canonicalUrl` contains the canonical URL of the primary document as defined by its `link rel=canonical` tag.
+- `window.context.clientId` contains a unique id that persistently the same for a given user and AMP origin site.
```
Comment (erwinmombay) [inline ads/README.md]: `that is persistently`

Comment (cramforce) [inline ads/README.md]: Done.

File: docs/create_page.md (hunk 3 above)
Comment (cramforce) [inline docs/create_page.md]: Apparently :)

File: src/cid.js (hunk 2 above)
Comment (cramforce) [inline src/cid.js]: Done, also added a doc string.

**Comment by erwinmombay** [comment]
LGTM.

just one last question, do we really need both `getElementService` and `getElementServiceIfAvailable` ? if i'm understanding correctly the former `asserts` and will throw while the latter yields `null` on the promise. Was wondering if we could consolidate on one pattern to reduce code. (ignore if we do have a good reason to have both)

**Comment by cramforce** [comment]
Its worth considering to consolidate the APIs, but I see that the `getElementService` will be more frequently used and it is essential that we have a really nice error message in case it fails. I don't want to make that message a responsibility of the caller.

**Comment by cramforce** [comment]
@avimehta @btownsend

File: src/service.js [hunk 5]
```diff
@@ -122,6 +123,26 @@ function getElementService_(win, id, providedByElement) {
...
 /**
+ * Same as getElementService but produces null if the given element is not
+ * actually available on the current page.
+ * @param {!Window} win
+ * @param {string} id of the service.
+ * @param {string} provideByElement Name of the custom element that provides
+ *     the implementation of this service.
+ * @return {!Promise<*>}
+ */
+export function getElementServiceIfAvailable(win, id, providedByElement) {
```
Comment (jridgewell) [inline src/service.js]: You're gonna need to rebase, this belongs in `src/custom-element.js` now.

Comment (cramforce) [inline src/service.js]: Rebased :)

Yaiihh, fun.

**Pull request merged**